fbgemm_gpu/test/uvm_test.py

```diff
@@ -19,13 +19,16 @@
 
 if open_source:
     # pyre-ignore[21]
-    from test_utils import gpu_available, gpu_unavailable
+    from test_utils import gpu_available, gpu_unavailable, skipIfRocm
 else:
     from fbgemm_gpu.test.test_utils import gpu_available, gpu_unavailable
 
 if gpu_available:
-    # pyre-ignore[21]
-    from fbgemm_gpu.uvm import cudaMemAdvise, cudaMemoryAdvise, cudaMemPrefetchAsync
+    if torch.version.hip:
+        # pyre-ignore[21]
+        from fbgemm_gpu.uvm import cudaMemAdvise, hipMemoryAdvise, cudaMemPrefetchAsync
+    else:
+        from fbgemm_gpu.uvm import cudaMemAdvise, cudaMemoryAdvise, cudaMemPrefetchAsync
```

Comment:
Why do we need to change only `cudaMemoryAdvise` to `hipMemoryAdvise`?
Can't we use the same symbols?

I understand that the hipification script changes those symbols, but this interface does not look consistent.

Reply:
Changes on hipify_torch needed for this issue.

Reply:
I truly understand why it happens and this is the easy fix, but I don't think this is the right interface we should provide since PyTorch and FBGEMM basically provides GPU features as "CUDA" if the corresponding features are provided by HIP/ROCm. Can you escape `hipify_python.py` somehow?

Reply:
Any updates on this @liligwu ?

Reply:
@shintaro-iwasaki @jianyuh I think this issue is cropping up because of the (unnecessary?) clash in the naming of the uvm enum `cudaMemoryAdvise` (https://github.com/pytorch/FBGEMM/blob/a70d14ca5f7f3354337d3ccfa5558c81fa550a00/fbgemm_gpu/src/cumem_utils.cu#L360) vs the CUDA-provided enum `cudaMemoryAdvise` (which has a HIP equivalent `hipMemoryAdvise` that needs to be used for ROCm). Are they required to be identically named? If not, I think the most straightforward solution here is to deduplicate the naming.

I agree that the python interface should be unified and hence the python enum should be the same for ROCm vs CUDA. That will also render changes like these unnecessary:
https://github.com/pytorch/FBGEMM/pull/1142/files#diff-063346a73da0b0a6d9cb2ba902a1635ae9f57beeda2df107648bf4724cf47962R85

Reply:
Thanks for your comment @jithunnair-amd. This PR provides the following for HIP:
```
from fbgemm_gpu.uvm import cudaMemAdvise, hipMemoryAdvise, cudaMemPrefetchAsync
```
This is not a right interface; the names are very inconsistent.


There should be four directions to implement this.

1. HIP provides both "hip" API and "cuda" alias
   - HIP provides both "hipMemAdvise", "hipMemoryAdvise", "hipMemPrefetchAsync", "cudaMemAdvise", "cudaMemoryAdvise", "cudaMemPrefetchAsync"
   - CUDA provides "cudaMemAdvise", "cudaMemoryAdvise", "cudaMemPrefetchAsync"

This sounds great to me if this is the proposal.

2. HIP provides "cuda"
   - HIP provides: "cudaMemAdvise", "cudaMemoryAdvise", "cudaMemPrefetchAsync"
   - CUDA provides "cudaMemAdvise", "cudaMemoryAdvise", "cudaMemPrefetchAsync"

This is what I proposed; if the hipification mechanism is somehow escaped, it should be easy.

3. HIP provides "hip" only
   - HIP provides: "hipMemAdvise", "hipMemoryAdvise", "hipMemPrefetchAsync"
   - CUDA provides "cudaMemAdvise", "cudaMemoryAdvise", "cudaMemPrefetchAsync"


We don't want this since the interface is not unified. We would appreciate option 1 for compatibility.

4. HIP mixes "hip" and "cuda"
   - HIP provides: "cudaMemAdvise", "hipMemoryAdvise", "cudaMemPrefetchAsync"
   - CUDA provides "cudaMemAdvise", "cudaMemoryAdvise", "cudaMemPrefetchAsync"

This is the current implementation. This interface is messy

---

I might be misunderstanding your point. Would you mind teaching us your proposal, @jithunnair-amd ?

Reply:
@shintaro-iwasaki Sure. What you want here is the separation of "church" (Python) and "state" (C++). In other words, the Python enum need not be exactly same as the C++ enum, however, IIUC, in this case, you're forcing that via the following code: https://github.com/pytorch/FBGEMM/blob/a70d14ca5f7f3354337d3ccfa5558c81fa550a00/fbgemm_gpu/fbgemm_gpu/uvm.py#L21

This creates a problem for hipify because you ideally want hipify to only modify the C++-side uses of CUDA APIs/enums, but *not* the python side. However, since you use the above code in conjunction with the following way to define the uvm enum (which then gets mimicked on the python side using the above code):
https://github.com/pytorch/FBGEMM/blob/a70d14ca5f7f3354337d3ccfa5558c81fa550a00/fbgemm_gpu/src/cumem_utils.cu#L361

Hipify will replace the identifier it thinks is a CUDA identifier here, because it doesn't know that it'll also be eventually used as the name of the enum on the Python side. That is why you end up having it show up here:
https://github.com/pytorch/FBGEMM/pull/1142/files#diff-063346a73da0b0a6d9cb2ba902a1635ae9f57beeda2df107648bf4724cf47962R27

**PROPOSAL**
Our proposal is the following:
1. Rename the python enums to be distinct from the CUDA C types eg.
`from fbgemm_gpu.uvm import gpuMemAdvise, gpuMemoryAdvise, gpuMemPrefetchAsync`
2. Use a different macro to map the python/uvm enum name to the CUDA type:
```
+#define FBGEMM_GPU_ENUM_ITEM2(x, y) \
+ { #x, y }
```
```
FBGEMM_GPU_ENUM_REGISTER_START(uvm, gpuMemoryAdvise){
 FBGEMM_GPU_ENUM_ITEM2(gpuMemAdviseSetReadMostly, cudaMemAdviseSetReadMostly),
```
3. This will make the python side quite clean as well:
```
assert gpuMemoryAdvise.gpuMemAdviseSetAccessedBy.value == 5
...
cudaMemAdvise(uvm_t, gpuMemoryAdvise.gpuMemAdviseSetAccessedBy)
```

This gives us a unified interface in python without hacking the hipify paradigm.

Reply:
We really want to keep the existing interface. It's really great if we can add another name (`gpuMemXXX`), but please keep the existing Python interface.

It's a hipification script issue, and I believe we can easily escape it.  For example, token concat (`##`) can avoid hipification without changing `hipify_python.py`. There should be a reasonably elegant way to keep the interface. To do so, the following you suggested would be a good idea.
```
#define FBGEMM_GPU_ENUM_ITEM2(x, y) \
 { #x, y }
```

Reply:
Hi @shintaro-iwasaki , we pushed a workaround for this Python API issue. Please review.

Reply:
Thanks. Not very clean (understandable though), but it looks good to me!